Add tests for reverse_lazy for #858

wgordon17 opened the pull request.

pedro-psb commented on tests_functional/django_example/settings.yaml:
It would be good to test if it is composable by chaining it with some other converter.

pedro-psb reviewed (commented):
It would be good to test if it is composable too by chaining it with some other converter.

wgordon17 commented:
@pedro-psb It _definitely doesn't seem composable. I made the following change in `settings.yaml

```yaml
  ADMIN_NAMESPACE: admin
  LOGIN_URL: "@reverse_lazy @format {this.ADMIN_NAMESPACE}:login"
```

And it blew up

```
TypeError: descriptor 'format' for 'str' objects doesn't apply to a 'Lazy' object
```

If I try to change the order and use `@format @reverse_lazy`, then the `@reverse_lazy` value just gets treated as raw text.

I'm open to any suggestions, otherwise we just call this a "noncomposable" parser? ¯\\\_(ツ)_/¯

pedro-psb commented:
@wgordon17 I'll try something out tomorrow, I don't know much about the converters' mechanism yet.

But yes, although it would be better to provide a composable example, it is already very useful to have a non-composable one.

wgordon17 commented:
Sounds good @pedro-psb, let me know if I can help. Ideally, I wouldn't have to use `Lazy` at all, but I can't seem to figure out a way past early-evalution. Since reversing Django routes relies on a fully configured project, but you can't configure the project without settings being fully setup, and the settings can't get setup because the route is trying to be reversed...and around and around we go 😉 

For me, https://github.com/dynaconf/dynaconf/blob/master/dynaconf/base.py#L894, was the offending line that was causing early value evaluation breaking the "laziness" of `reverse_lazy`. But wrapping it in a `Lazy` object seemed to fix the early evaluation issue. Perhaps there's a _different_ way to wrap the parser/casting to still "act" like a `str` while still delaying evaluation ¯\\\_(ツ)_/¯

pedro-psb commented:
@wgordon17 I've found out that the caller of the converters (`get_converters`) optionally passes the `box_settings` to the converter functions.
Then, you can use it to evaluate the `value` if it got there as a Lazy object (because `@format` needs to be Lazy too).

```python
def reverse_lazy_fun(value, box_settings=None):
    if isinstance(value, Lazy):
        # value got here as a Lazy object, needs evaluation
        return Lazy(value(box_settings), casting=reverse_lazy)
    else:
        # value is good to go
        return Lazy(value, casting=reverse_lazy)


parse_conf.converters["@reverse_lazy"] = reverse_lazy_fun
```

rochacbruno commented:
@pedro-psb I think we must add a builtin helper to register new converters then we abstract this logic

```py
def reverse_lazy_fun(value, box_settings=None):
    if isinstance(value, Lazy):
        # value got here as a Lazy object, needs evaluation
        return Lazy(value(box_settings), casting=reverse_lazy)
    else:
        # value is good to go
        return Lazy(value, casting=reverse_lazy)
```

In something better to use like

```py
settings.dynaconf.add_converter("reverse_lazy", callable, combo=True)
```

pedro-psb commented:
@rochacbruno Yes, I think this would be good

rochacbruno commented:
This is what I was playing with @pedro-psb 
I will add to a PR later today, need to test with more scenarios.

Also I am not sure if keep it as a standalone function or bind it to the settings as a method

![test_formater_helper](https://user-images.githubusercontent.com/458654/228603792-121e3eca-c4dd-4d49-b5a5-06c475f68067.png)

wgordon17 commented:
@rochacbruno I was definitely not being elegant, but I just threw this together as well https://github.com/dynaconf/dynaconf/pull/904/files#diff-81216dc760c0d7472d00bab2b35a37d32985b9b0e07ffa9877ed2bbb664373c1

rochacbruno commented:
@wgordon17 do we really need to pass `box_settings` ? maybe your implementation is better

wgordon17 commented:
@rochacbruno Heh, I really don't know 🙈 I only stole it from @pedro-psb's suggestion. I think you're the SME on the backend internals

rochacbruno commented on dynaconf/utils/parse_conf.py:
```diff
@@ -276,6 +276,20 @@ def get_converter(converter_key, value, box_settings):
 
 
+def add_converter(converter_key, casting_callable):
+    global converters
```
No need to be global here because `converters` is mutable

rochacbruno commented on dynaconf/utils/parse_conf.py:
```diff
@@ -276,6 +276,20 @@ def get_converter(converter_key, value, box_settings):
+            return Lazy(value(box_settings), casting=casting_callable)
+        else:
+            # value is good to go
+            return Lazy(value, casting=casting_callable)
```
Would be good to set a `formatter` here, the default is `str.format` which works fine but we can avoid that with

```py
def noop_formatter_factory(converter_key):
    """Builds a new noop formatter and sets its token"""
    return BaseFormatter(lambda x, **_: x, converter_key)
```

and

```suggestion
            return Lazy(
                value, 
                casting=casting_callable, 
                formatter=noop_formatter_factory(f"@{converter_key}")
            )
```

rochacbruno commented on dynaconf/utils/parse_conf.py:
```diff
@@ -276,6 +276,20 @@ def get_converter(converter_key, value, box_settings):
+def add_converter(converter_key, casting_callable):
+    global converters
+
+    def converter_callable(value, box_settings=None):
```
We can use functools.wraps to make this generated function to inherit name and docstring from the passed function. https://docs.python.org/3/library/functools.html#functools.wraps

```suggestion
    @functools.wraps(casting_callable)
    def converter_callable(value, box_settings=None):
```

pedro-psb commented:
I think `add_converter` should be a module function, as it is modifying a module variable.

It could live in this same file (`parse_conf.py`), but as it is a internals module, it would be better to have a public `utils` or similar namespace just for users, so their autocompletion don't get crowded with plumbing stuff.

rochacbruno commented on dynaconf/utils/parse_conf.py:
```diff
@@ -276,6 +276,20 @@ def get_converter(converter_key, value, box_settings):
     return converted_value
 
 
+def add_converter(converter_key, casting_callable):
+    global converters
+
+    def converter_callable(value, box_settings=None):
+        if isinstance(value, Lazy):
+            # value got here as a Lazy object, needs evaluation
+            return Lazy(value(box_settings), casting=casting_callable)
+        else:
+            # value is good to go
+            return Lazy(value, casting=casting_callable)
+
+    converters[f"@{converter_key}"] = converter_callable
```
Just a suggestion, it is in practice the same, so any of the options might work

```suggestion
def add_converter(converter_key, func):
    """Adds a new converter to the converters dict"""
    if not converter_key.startswith("@"):
        converter_key = f"@{converter_key}"
    converters[converter_key] = wraps(func)(
        lambda value: value.set_casting(func) if isinstance(value, Lazy)
        else Lazy(
            value,
            casting=func,
            formatter=BaseFormatter(lambda x, **_: x, converter_key)
        )
    )

```

rochacbruno commented:
then on `dynaconf/__init__.py` we can do

```py
from dynaconf.utils.parse_conf import add_converter

__all__ = [
    ...
   "add_converter"
]
```

then the usage would be

```py
from dynaconf import Dynaconf, add_converter

settings = Dynaconf()
add_converter("key", function)
```

wgordon17 commented:
Rebased and refactored with suggestions, thanks so much @pedro-psb @rochacbruno! Also included a quick change to the docs. Once you're good with this, I can squash all the commits down (or GitHub can do it ¯\\\_(ツ)_/¯)
